Cloud Application Manager/Automating Deployments/running-code-outside-an-instance.md

```diff
@@ -0,0 +1,99 @@
+{{{ "title": "Running Code outside an Instance with Task Variables",
+"date": "26-04-2018",
+"author": "",
+"attachments": [],
+"contentIsHTML": false
+}}}
+
+The event scripts are run, by definition, on the instance they belong to. You may have a need, though, to run some code outside the instance's operating system, either because the code must run before the instance has been provisioned or because it has already been terminated. In these cases you may use task variables.
+
+### Basics
+
+Task variables define code and an associated event. When a box with a task variable is deployed, CAM will run this code within the instance's lifecyle, at a moment indicated by the task variable's event:
+
+* `deploy`: as the first step after you click *Deploy*, before any provisioning is run on the provider.
+* `remove`: as the last step once the instance has been terminated on the provider. 
+
+Although not run within the instance itself, task variables behave similarly to event scripts:
+
+* Jinja templating [syntax](syntax-for-variables.md) is available, so you may access variables from within the script.
+* `elasticbox set` and `elasticbox config` [commands](cloud-application-manager-commands.md) are available.
+* When using box composition, task variables of child boxes are run first during `deploy` and `provisioned` events, whereas they are run last during `remove` events.
+* A failure during a task variable execution will result in an unavailable state. A failure may be triggered from within the script with an exit code different from 0 or may occur if the script took too long to complete and was forcibly stopped.
+* Output from the script will appear in the instance activiy log.
+
+### Languages
+
+The environment supports bash, python and nodejs script. By default, however, all scripts should start as shell scripts. You may then embed code in other languages and execute it from within the shell script.
+
+### Availability and visibility
+
+Task variables are only available for public, versioned, verified boxes. Because of this verification process, there is no Task Variable option in the code section of a box: you should create a File variable and then send a publish request. The variables will be marked as *private* and will not appear in the lifecycle editor. 
```

Comment:
Please, also mention here that in the publish request the requester should mention in comments the file var that wants to be promoted to task var, and the event that he wants the var to be executed.